Exiled.CustomRoles/API/Features/CustomRole.cs

```diff
@@ -0,0 +1,445 @@
+        /// </summary>
+        /// <param name="player">The <see cref="Player"/> to remove the role from.</param>
+        public virtual void RemoveRole(Player player)
+        {
+            TrackedPlayers.Remove(player);
+            if (RemovalKillsPlayer)
+                player.Role = RoleType.Spectator;
+            foreach (CustomAbility ability in CustomAbilities)
+            {
+                ability.RemoveAbility(player);
+            }
+
+            RoleRemoved(player);
+        }
+
+        /// <summary>
+        /// Tries to add an item to the player's inventory by name.
+        /// </summary>
+        /// <param name="player">The <see cref="Player"/> to try giving the item to.</param>
+        /// <param name="itemName">The name of the item to try adding.</param>
+        /// <returns>Whether or not the item was able to be added.</returns>
+        protected bool TryAddItem(Player player, string itemName)
+        {
+            if (CustomItem.TryGet(itemName, out CustomItem customItem))
+            {
+                customItem.Give(player);
+
+                return true;
+            }
+
+            if (Enum.TryParse(itemName, out ItemType type))
+            {
+                if (type.IsAmmo())
+                    player.Ammo[type] = 100;
+                else
+                    player.AddItem(type);
+
+                return true;
+            }
+
+            Log.Warn($"{Name}: {nameof(TryAddItem)}: {itemName} is not a valid ItemType or Custom Item name.");
+
+            return false;
+        }
+
+        /// <summary>
+        /// Gets a random <see cref="Vector3"/> from <see cref="SpawnProperties"/>.
+        /// </summary>
+        /// <returns>The chosen spawn location.</returns>
+        protected Vector3 GetSpawnPosition()
+        {
+            if (SpawnProperties == null || SpawnProperties.Count() == 0)
+                return Vector3.zero;
+
+            if (SpawnProperties.StaticSpawnPoints.Count > 0)
+            {
+                foreach ((float chance, Vector3 pos) in SpawnProperties.StaticSpawnPoints)
+                {
+                    int r = Loader.Random.Next(100);
+                    if (r <= chance)
+                        return pos;
+                }
+            }
+
+            if (SpawnProperties.DynamicSpawnPoints.Count > 0)
+            {
+                foreach ((float chance, Vector3 pos) in SpawnProperties.DynamicSpawnPoints)
+                {
+                    int r = Loader.Random.Next(100);
+                    if (r <= chance)
+                        return pos;
+                }
+            }
+
+            if (SpawnProperties.RoleSpawnPoints.Count > 0)
+            {
+                foreach ((float chance, Vector3 pos) in SpawnProperties.RoleSpawnPoints)
+                {
+                    int r = Loader.Random.Next(100);
+                    if (r <= chance)
+                        return pos;
+                }
+            }
```

Comment:
Joker please for the love of god just make this a method. Also can't it just not pick a role sometimes?

Reply:
I'm not actually sure that's possible, since each of those 3 values are different types, I'd have to re-build the list of them as generic spawn points for a single method to exist to iterate them and pick one at random.

Reply:
Looks like C# is weird and this cannot be a method, so I guess it's fine.